### PR TITLE
Add automatic differentiation on the Lie group tangent space

### PR DESCRIPTION
**Pull request by fantaosha**
## Motivation and Context

Add support for the automatic Jacobian and gradient computation on the Lie group tangent space.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by fantaosha** [comment]
The automatic differentiation on the Lie group tangent space works for all the Lie groups in `theseus`. Maybe we can test some examples without providing any `Jacobians` manually.

**Review by mhmukadam — APPROVED** [review]
Looks great, nice work!

File: theseus/geometry/tests/test_so2.py [hunk 1]
```diff
@@ -108,3 +108,44 @@ def test_copy():
     rng.manual_seed(0)
     so2 = _create_random_so2(1, rng)
     check_copy_var(so2)
+
+
+def test_projection():
```
Comment (fantaosha) [inline theseus/geometry/tests/test_so2.py:113]: This is used to test the correctness of the automatic Jacobean calculation.

File: theseus/geometry/tests/test_se3.py [hunk 2]
```diff
@@ -185,3 +185,46 @@ def test_transform_from_and_to():
             )
             assert torch.allclose(jacobians_from[0], expected_jac[0])
             assert torch.allclose(jacobians_from[1], expected_jac[1])
+
+
+def test_projection():
```
Comment (fantaosha) [inline theseus/geometry/tests/test_se3.py:190]: This is used to test automatic jacobian calculation for `SE(3)`

File: examples/se2_inverse.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+from theseus.geometry.tests.test_se2 import create_random_se2
+
+rng = torch.Generator()
+rng.manual_seed(0)
+x1 = create_random_se2(1, rng)
+x2 = create_random_se2(1, rng)
+
+
+def run(x1: LieGroup, x2: LieGroup, num_iters=10, use_lie_tangent=True):
```
Comment (luisenp) [inline examples/se2_inverse.py:23]: Given that this is in the examples folder, let's add some text at the beginning explaining what this example is showing.

File: examples/se2_inverse.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    x1.data = LieGroupTensor(x1)
```
Comment (luisenp) [inline examples/se2_inverse.py:24]: Is the change that is going to make `LieGroup.data` be of type `LieGroupTensor` going to be part of this PR?

File: theseus/geometry/tests/test_se2.py [hunk 5]
```diff
@@ -120,3 +120,29 @@ def test_theta_jacobian():
...
+                return th.SE2(data=g).transform_to(p).data.sum(dim=0)
+
+            jac = torch.autograd.functional.jacobian(
+                transform_to_sum, (se2.data, point.data)
+            )
+
+            actual = [
+                se2.project(jac[0]).transpose(0, 1),
+                point.project(jac[1]).transpose(0, 1),
+            ]
```
[the remark this answers is not in the document]
Comment (luisenp) [reply theseus/geometry/tests/test_se2.py:144]: I think this is just needed for the test to compare that `project(euclidean_grad)` gives the same result as `tangent_grad`. BTW, @fantaosha, why is the `transpose` needed? Isn't dim=0 a batch dimension? 

Also, a few days ago we mentioned that `AutodiffCostFunction` will always call `project` on the computed jacobians before [returning](https://github.com/facebookresearch/theseus/blob/main/theseus/core/cost_function.py#L178). Should this be part of this PR? (this is related to Mustafa's question)

**Review by luisenp — COMMENTED** [review]
Left some questions to address before merging.

File: examples/se2_inverse.py (hunk 3 above)
Comment (mhmukadam) [reply examples/se2_inverse.py:23]: Also update the readme in that folder accordingly.

**Comment by fantaosha** [comment]
Please refer to #98 for enhanced `LieGroup.autograd`.

File: examples/se2_inverse.py (hunk 4 above)
Comment (fantaosha) [reply examples/se2_inverse.py:24]: No. This will cause more issues than I previously thought.

File: examples/se2_inverse.py (hunk 3 above)
Comment (fantaosha) [reply examples/se2_inverse.py:23]: Will do.

File: examples/se2_inverse.py (hunk 4 above)
Comment (luisenp) [reply examples/se2_inverse.py:24]: Let's merge without this. I'll look at this part next week and follow up with you if I run into some problems.

**Review by mhmukadam — APPROVED** [review]

**Pull request merged**